Document Roslyn Diagnostics Security Analyzer

BillWagner opened the pull request:
Fixes #1669 

Note that the configuration instructions in the blog post referenced in issue #1669 are not longer needed.

This PR is dependent on #3908

BillWagner committed:
first draft

Ready for a proofread.

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+# The .NET Security Analyzers
+
+You can use the .NET Security Analyzers to find potential security risks
+in your code. These analyzers will find potential issues, and suggest
```
no comma after "issues"

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+in your code. These analyzers will find potential issues, and suggest
+fixes to those issues.
+
+The analyzers run interactively in Visual Studio, as you write your code,
```
no comma either before as and after code.

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+or as part of a CI build. You'll want to add the analyzer to your project as
+early as possible in your development. The sooner you find any potential issues
+in your code, the easier it will be to fix. However, you can add it at any time
+in the development cycle. It will find any issues with the existing code, and
```
no comma before and

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+in the development cycle. It will find any issues with the existing code, and
+will start warning about new issues as you keep developing.
+
+## Installing and configuring the .NET Security Analyzer
```
Analyzer --> Analyzers

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+
+The .NET Security Analyzers must be installed as a NuGet package on every
+project where you want them to run. Only one developer needs to add them
+to the project. The analyzer package is a project dependency, and will run
```
no comma after dependency

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+The .NET Security Analyzers are part of the [Microsoft.DesktopAnalyzers](https://www.nuget.org/packages/Desktop.Analyzers/)
+NuGet package. This package provides a smaller set of analyzers
+that includes the security analyzers. In most cases, you'll want
+the [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) NuGet package. 
```
Are the SecurityAnalyzers also included in FxCopAnalyzers? In any case, the relationship is unclear.

BillWagner replied:
I re-wrote it to be more clear.

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+"Fx Cop Analyzers". Install the latest stable version in all projects in your
+solution.
+
+Once the NuGet package is installed run a build, and you'll see a list of
```
comma after installed
no comma after build

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+solution.
+
+Once the NuGet package is installed run a build, and you'll see a list of
+warnings for the three security related issues found by these analyzers:
```
what is the analyzer running against to produce these warnings? Does it come with a project that produces the warnings?

BillWagner replied:
I re-wrote it to be more clear.

rpetrusha commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,61 @@
+
+Code that triggers any of these analyzers generates a warning by default.
+
+## Roslyn based analyzers vs. Code Analysis libraries
```
Here and throughout, Roslyn based --> Roslyn-based

rpetrusha reviewed (changes requested):
Left some questions and comments, @BillWagner.

BillWagner committed:
respond to feedback

BillWagner commented:
I updated this based on your feedback @rpetrusha

BillWagner commented:
Do you want to provide a tech review @mavasani

mavasani commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,62 @@
+to the project. The analyzer package is a project dependency and will run
+on every developer's machine once they have the updated solution.
+
+The .NET Security Analyzers are delivered in the [Microsoft.DesktopAnalyzers](https://www.nuget.org/packages/Desktop.Analyzers/)
```
Desktop analyzers package is no longer maintained, and we should not recommend anyone to install them. We have refactored the analyzers into following packages:

**Core analyzer packages**
1. [Microsoft.CodeQuality.Analyzers](https://www.nuget.org/packages/Microsoft.CodeQuality.Analyzers): Common code quality rules, which are not specific to any .net framework or APIs. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeQuality.Analyzers/Microsoft.CodeQuality.Analyzers.md)
2. [Microsoft.NetCore.Analyzers](https://www.nuget.org/packages/Microsoft.NetCore.Analyzers): Analyzers specific to netcore APIs. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.NetCore.Analyzers/Microsoft.NetCore.Analyzers.md)
3. [Microsoft.NetFramework.Analyzers](https://www.nuget.org/packages/Microsoft.NetFramework.Analyzers): Analyzers specific to netfx APIs. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.NetFramework.Analyzers/Microsoft.NetFramework.Analyzers.md)
4. [Text.Analyzers](https://www.nuget.org/packages/Text.Analyzers): Analyzers specific to text APIs, e.g. for analyzer and code fix authors. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Text.Analyzers/Text.Analyzers.md)
5. [Microsoft.CodeAnalysis.Analyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.Analyzers): Analyzers specific to usage of Roslyn/Microsoft.CodeAnalysis APIs, e.g. for analyzer and code fix authors. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.md)

**Aggregate analyzer packages**
1. [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers/): Analyzers aimed at replacement of legacy FXCop rules. This package is an aggregate of the following analyzer packages:
   1. Microsoft.CodeQuality.Analyzers
   2. Microsoft.NetCore.Analyzers
   3. Microsoft.NetFramework.Analyzers
   4. Text.Analyzers

BillWagner replied:
I'll expand this PR to cover NetFramework.Analyzers.  I've opened the following new issues for the others that are part of the FxCopAnalyzers package:

#4214
#4215
#4216

Note that the CodeAnalysis package will be covered as part of the Roslyn API documentation.

mavasani commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,62 @@
+
+Once the NuGet package is installed, build your solution. The analyzers examine
+the code in your solution and provide you with a list of
+warnings for the three security related issues:
```
> three security related issues

These rules have been moved to `Microsoft.NetFramework.Analyzers`

mavasani commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,62 @@
+The user experience for Roslyn-based analyzers is different than that of the
+Code Analysis libraries like the older versions of FxCop and the Security analysis
+tools.  You don't need to explicitly run the Roslyn-based analyzers. There's no need to use the "Run Code Analysis"
+menu items on the "Analyze" menu in Visual Studio. Roslyn-based analyzers run asychronously as you work.
```
You probably want to also mention that doing a "Build" in Visual Studio or command line will also cause the analyzers to run and report diagnostics, just like compiler diagnostics. In fact, analyzers are essentially plugins that C# and VB compilers invoke during build and live analysis, and then it aggregates and outputs the compiler and analyzer diagnostics.

BillWagner replied:
Moved to the index.md file in the analyzers directory as this applies to all Roslyn-based analyzers.

BillWagner commented:
Adding @richlander (wrote #1669) and @OliaG (wrote #3908)

Reading @mavasani 's comments, I think the focus here is wrong based on the direction and the packaging. The analyzers that were part of the security set are now in the NetFramework and NetCore packages, where applicable.  That means this article would need an update to direct users to different packages for Core and Framework projects.

I'll propose that a better way to fix #1669 is to write an article that discusses using the first 4 analyzers @mavasani mentions in his comment above, and the aggregate FxCop analzyer package.

I could rework what's here (including changing the file name) and have it done in the next sprint.

Upvote if you agree, downvote with suggestions if you disagree.

BillWagner committed:
interim checkin

DesktopAnalyzers project is not being maintained, so reference NetFramework Analyzers instead.

More larger scale changes may be coming.

BillWagner commented:
Adding the [WIP] label because some larger changes may be coming.

richlander commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,63 @@
+Desktop.Analyzers package as well as other analyzers that
+help your team enforce the rules recommended by FxCop.
+
+To install it, right-click on the project, and select "Manage Dependencies".
```
Is there a separate topic you can point to that walks through the details of NuGet package installation more generally?

BillWagner replied:
I added this to a new index.md file in the /standard/analyzers folder. That will be applicable to the changes in #3908

richlander commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,63 @@
+"Fx Cop Analyzers". Install the latest stable version in all projects in your
+solution.
+
+Once the NuGet package is installed, build your solution. The analyzers examine
```
Maybe this is a start to a "Using the .NET Security Analyzers" section?

richlander commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,63 @@
+- [CA5350](~/visualstudio/code-quality/ca5350-do-not-use-weak-cryptographic-algorithms): Do not use weak cryptographic algorithms
+- [CA5351](~/visualstudio/code-quality/ca5351-do-not-use-broken-cryptographic-algorithms): Do not use broken cryptographic algorithms
+
+Code that triggers any of these analyzers generates a warning by default.
```
Seems like a good opportunity to a screenshot

richlander commented on docs/standard/security-analyzer.md:
```diff
@@ -0,0 +1,63 @@
+
+Code that triggers any of these analyzers generates a warning by default.
+
+## Roslyn-based analyzers vs. Code Analysis libraries
```
This section threw me off a bit.  I didn't realize at first that this was a comparison and not more info about the main topic.  I don't have specific suggestions on how to solve that.

BillWagner replied:
I moved it to a note in the index.md file that I added. I hope that sets it as a "sidebar" to the main topic.

BillWagner committed:
fix links

richlander commented:
The other option is to describe the general capability (which is close to what you have here) and then link to platform-specific docs for the specific instructions for using the analyzers. The benefit of this approach would be two-folder: separation of concerns and you'd get readers in both the standard and non-standard docs. Something to consider.

mairaw commented:
Haven’t reviewed this yet, but you might wanna put this topic in the analyzers folder we’re creating in the dependent PR. We also need to create an index page either here or there or on a different PR.

BillWagner committed:
reorganize folders

More analyzers means we should put them in a folder, and add an index.md to describe common tasks

BillWagner committed:
add index for landing page.

BillWagner committed:
Updated, except for screen shot

And a nice example.

BillWagner committed:
switching machine

Pushing to switch machines.

BillWagner committed:
add screenshots

Using the analyzer now contains screenshots and shows how it reports errors.

BillWagner commented:
I updated and responded to all the feedback. This is ready for another review.

rpetrusha reviewed (changes requested):
This topic looks really good, @BillWagner, although I've left a number of comments.

BillWagner commented on docs/standard/analyzers/framework-analyzer.md:
```diff
@@ -0,0 +1,154 @@
+
+![interactive report of issues found by the framework analyzer](./media/framework-analyzers-1.png)
+
+## images to add:
```
oops.  :blush:

BillWagner committed:
respond to feedback.

BillWagner commented:
Thanks @rpetrusha I've responded to all the comments.

rpetrusha commented on docs/standard/analyzers/framework-analyzer.md:
```diff
@@ -0,0 +1,158 @@
+Framework.Analyzers package as well as the following analyzers:
+- [Microsoft.CodeQuality.Analyzers](https://www.nuget.org/packages/Microsoft.CodeQuality.Analyzers): Provides general guidance and guidance for .NET Standard APIs
+- [Microsoft.NetCore.Analyzers](https://www.nuget.org/packages/Microsoft.NetCore.Analyzers): Provides analyzers specific to .NET Core APIs.
+- [Text.Analyzers](https://www.nuget.org/packages/Text.Analyzers): Provides guidance for text included as code including comments.
```
nit: there should be a comma after "code"

rpetrusha reviewed (approved):
Great job on the topic and the revision, @BillWagner. I've left one nit, which you can choose to address or ignore.

BillWagner committed:
one final grammar fix
